agent/credentials/instancecreds/instancecreds.go

```diff
@@ -38,7 +39,25 @@ func GetCredentials() *credentials.Credentials {
 	mu.Lock()
 	if credentialChain == nil {
 		credProviders := defaults.CredProviders(defaults.Config(), defaults.Handlers())
-		credProviders = append(credProviders, providers.NewRotatingSharedCredentialsProvider())
+		/*
+			The default credential chain provided by the SDK includes:
+			* EnvProvider
+			* SharedCredentialsProvider
+			* RemoteCredProvider
+
+			In the case of ECS-A on Windows, the `SharedCredentialsProvider` takes
+			precedence over the `RotatingSharedCredentialsProvider` and this results
+			in the credentials not being refreshed. To mitigate this issue, we will
+			use the environment variable `ECS_EXTERNAL` to reorder the credential
+			chain and ensure that `RotatingSharedCredentialsProvider` takes precedence
+			over the `SharedCredentialsProvider`.
+
+		*/
+		if _, ok := os.LookupEnv("ECS_EXTERNAL"); ok {
```

Comment:
Do we need to do this only for Windows?

Reply:
What is the case any of the other providers are used? Should we just be putting the rotating creds as the only one in the list?

Reply:
Currently, this appears to be broken only for ECS-A Windows. 

In certain cases, customers might want to use the `EnvProvider` and I think that it should always be the first provider in the chain.

Reply:
Ok sounds good.

Reply:
This will return true if ECS_EXTERNAL=false is set.

But you shouldn't be looking up env vars here anyways, you should use the config struct for this, like we do here: https://github.com/aws/amazon-ecs-agent/blob/master/agent/app/agent.go#L176-L180

Reply:
alternatively we could only do this reordering for windows, since on linux we run inside a container and don't have an issue with credential file collision

Reply:
The config struct cannot be used directly as it would result in an import cycle. 

However, we could pass in a boolean flag like `isExternal` to this method and update all its references within the repo. 

Ex:
```
* https://github.com/aws/amazon-ecs-agent/blob/225bc3a556bd2d1759ab27b23f54e7e68086c9f0/agent/ecr/factory.go#L78
* https://github.com/aws/amazon-ecs-agent/blob/225bc3a556bd2d1759ab27b23f54e7e68086c9f0/agent/ec2/ec2_client.go#L56
* https://github.com/aws/amazon-ecs-agent/blob/ed81f6b41241ab8ebba1bcec780352c58bf77bd0/agent/ec2/ec2_metadata_client.go#L97
* https://github.com/aws/amazon-ecs-agent/blob/f5ac255da315b39612ad95361247f74eb3db93ae/agent/app/agent.go#L228
```

Reply:
Do you recommend that the Linux credential chain remain unchanged for ECS-A?